File: backend/app/models/document.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from sqlalchemy import Column, ForeignKey, String, Table, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ..core.database import Base

# Association table
document_tags = Table(
    "document_tags",
    Base.metadata,
    Column("document_id", ForeignKey("documents.id"), primary_key=True),
    Column("tag_id", ForeignKey("tags.id"), primary_key=True),
)

DocumentTag = document_tags


class Document(Base):
    __tablename__ = "documents"

    id: Mapped[int] = mapped_column(primary_key=True)
    filename: Mapped[str] = mapped_column(String(255), nullable=False)  # stored filename
    original_name: Mapped[str] = mapped_column(String(255), nullable=False)
    content_text: Mapped[Optional[str]] = mapped_column(Text, nullable=True)  # extracted text
    file_type: Mapped[str] = mapped_column(String(20), nullable=False)  # pdf, docx, etc
    file_size: Mapped[int] = mapped_column(nullable=False)  # bytes
    folder_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("folders.id"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        default=datetime.utcnow, onupdate=datetime.utcnow
    )

    folder = relationship("Folder", back_populates="documents")
    tags = relationship("Tag", secondary=document_tags, back_populates="documents")

File: frontend/src/types/document.ts
export type Document = {
  id: number;
  filename: string;
  original_name: string;
  file_type: string;
  file_size: number;
  folder_id: number | null;
  created_at: string;
};

export type DocumentDetail = Document & {
  content_text: string | null;
};

export type DocumentListResponse = {
  items: Document[];
  total: number;
};

export const FILE_TYPE_OPTIONS = [
  { label: 'PDF', value: 'pdf' },
  { label: 'DOC', value: 'doc' },
  { label: 'DOCX', value: 'docx' },
  { label: 'Markdown', value: 'md' },
  { label: 'XLS', value: 'xls' },
  { label: 'XLSX', value: 'xlsx' }
] as const;

export function formatFileSize(bytes: number): string {
  if (!Number.isFinite(bytes) || bytes < 0) return '-';
  if (bytes < 1024) return `${bytes} B`;
  const kb = bytes / 1024;
  if (kb < 1024) return `${kb.toFixed(1)} KB`;
  const mb = kb / 1024;
  if (mb < 1024) return `${mb.toFixed(1)} MB`;
  const gb = mb / 1024;
  return `${gb.toFixed(1)} GB`;
}

